### PR TITLE
feat: add `UnescapeMappingTemplate` to state machine `Api` event

### PR DESCRIPTION
**Pull request by hoffa**
### Issue #, if available

https://github.com/aws/serverless-application-model/issues/1895

### Description of changes

Adds `UnescapeMappingTemplate` boolean to [state machine `Api` event](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-property-statemachine-statemachineapi.html), which unescapes singles quotes in state machine input escaped by `escapeJavaScript` (which isn't valid JSON).

The final mapping template isn't JSON by design, so can't use `json.dumps()`. For example:

```json
{"input": "$util.escapeJavaScript($input.json('$')).replaceAll("\\'","'")", "stateMachineArn": "arn:aws:states:us-east-1:792766875239:stateMachine:Post-n7ZKNljUJsZZ"}
```

Originally done in https://github.com/aws/serverless-application-model/pull/2253, but using an opt-in property instead to avoid large-scale redeployments.

### Description of how you validated changes

```bash
pytest integration --no-cov -k test_state_machine_with_api_single_quotes_input -s
pytest --no-cov samtranslator -n auto tests/translator/test_translator.py -k state_machine_with_api
```

### Checklist

- [x] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [ ] Add/update [unit tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions) using:
    - [ ] Correct values
    - [ ] Bad/wrong values (None, empty, wrong type, length, etc.)
    - [ ] [Intrinsic Functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html)
- [x] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)
- [ ] Update documentation
- [ ] Verify transformed template deploys and application functions as expected
- [ ] Do these changes include any template validations?
    - [ ] Did the newly validated properties support intrinsics prior to adding the validations? (If unsure, please review [Intrinsic Functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html) before proceeding).
        - [ ] Does the pull request ensure that intrinsics remain functional with the new validations?

### Examples?

Please reach out in the comments, if you want to add an example. Examples will be 
added to `sam init` through https://github.com/awslabs/aws-sam-cli-app-templates/

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by GavinZZ — APPROVED** [review]

**Comment by xazhao** [comment]
Do we have doc update for this?

File: integration/single/test_basic_api.py [hunk 1]
```diff
@@ -113,3 +115,21 @@ def test_basic_api_with_tags(self):
         self.assertIsNotNone(stage)
         self.assertEqual(stage["tags"]["TagKey1"], "TagValue1")
         self.assertEqual(stage["tags"]["TagKey2"], "")
+
+    def test_state_machine_with_api_single_quotes_input(self):
```
Comment (xazhao) [inline integration/single/test_basic_api.py:119]: Another test without `UnescapeMappingTemplate` to show it's not desired?

File: integration/helpers/base_test.py [hunk 2]
```diff
@@ -560,6 +560,17 @@ def verify_options_request(self, url, expected_status_code, headers=None):
...
         return response
 
+    def verify_post_request(self, url: str, body_obj, expected_status_code: int):
+        """Return response to POST request and verify matches expected status code."""
+        response = self.do_post_request(url, body_obj)
+        if response.status_code != expected_status_code:
+            raise StatusCodeError(
+                "Request to {} failed with status: {}, expected status: {}".format(
+                    url, response.status_code, expected_status_code
+                )
```
Comment (aahung) [inline integration/helpers/base_test.py:570]: nit: we should switch to use f-string now like we did for new code. Fine for testing code

**Review by aahung — APPROVED** [review]

File: integration/single/test_basic_api.py (hunk 1 above)
Comment (hoffa) [reply integration/single/test_basic_api.py:119]: It's always desired AFAICS, but have to do as opt-in property to avoid large-scale redeployments.

**Review by xazhao — APPROVED** [review]

**Pull request merged**